src/AggregateFunctions/AggregateFunctionSequenceNextNode.h

```diff
@@ -0,0 +1,462 @@
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypeDateTime.h>
+
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnVector.h>
+#include <Columns/ColumnNullable.h>
+
+#include <Common/ArenaAllocator.h>
+#include <Common/assert_cast.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <AggregateFunctions/AggregateFunctionNull.h>
+
+#include <type_traits>
+
+
+namespace DB
+{
+
+
+/// NodeBase used to implement a linked list for storage of SequenceNextNodeImpl
+template <typename Node>
+struct NodeBase
+{
+    UInt64 size; // size of payload
+
+    DataTypeDateTime::FieldType event_time;
+    UInt32 events_bitset; // Bitsets of UInt32 are easy to compare. (< operator on bitsets)
+
+    char * data() { return reinterpret_cast<char *>(this) + sizeof(Node); }
+
+    const char * data() const { return reinterpret_cast<const char *>(this) + sizeof(Node); }
+
+    Node * clone(Arena * arena) const
+    {
+        return reinterpret_cast<Node *>(
+            const_cast<char *>(arena->alignedInsert(reinterpret_cast<const char *>(this), sizeof(Node) + size, alignof(Node))));
+    }
+
+    void write(WriteBuffer & buf) const
+    {
+        writeVarUInt(size, buf);
+        buf.write(data(), size);
+
+        writeBinary(event_time, buf);
+        writeBinary(events_bitset, buf);
+    }
+
+    static Node * read(ReadBuffer & buf, Arena * arena)
+    {
+        UInt64 size;
+        readVarUInt(size, buf);
+
+        Node * node = reinterpret_cast<Node *>(arena->alignedAlloc(sizeof(Node) + size, alignof(Node)));
+        node->size = size;
+        buf.read(node->data(), size);
+
+        readBinary(node->event_time, buf);
+        readBinary(node->events_bitset, buf);
+
+        return node;
+    }
+};
+
+/// It stores String, timestamp, bitset of matched events.
+struct NodeString : public NodeBase<NodeString>
+{
+    using Node = NodeString;
+
+    static Node * allocate(const IColumn & column, size_t row_num, Arena * arena)
+    {
+        StringRef string = assert_cast<const ColumnString &>(column).getDataAt(row_num);
+
+        Node * node = reinterpret_cast<Node *>(arena->alignedAlloc(sizeof(Node) + string.size, alignof(Node)));
+        node->size = string.size;
+        memcpy(node->data(), string.data, string.size);
+
+        return node;
+    }
+
+    void insertInto(IColumn & column)
+    {
+        assert_cast<ColumnString &>(column).insertData(data(), size);
+    }
+};
+
+/// TODO : Expends SequenceNextNodeGeneralData to support other types
+template <typename T, typename Node, bool Descending>
+struct SequenceNextNodeGeneralData
+{
+    using Allocator = MixedAlignedArenaAllocator<alignof(Node *), 4096>;
+    using Array = PODArray<Node *, 32, Allocator>;
+
+    Array value;
+    bool sorted = false;
+
+    struct Comparator final
+    {
+        bool operator()(const Node * lhs, const Node * rhs) const
+        {
+            if constexpr (Descending)
```

Comment:
Seems that there are little in-determinism, result may vary if there are several results with same time:
```
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:01', 1, 'A');
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:01', 1, 'B');
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:02', 1, 'C');
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:02', 1, 'X');

SELECT sequenceNextNode(0)(dt, action, action = 'A', action = 'B') AS next_node  FROM test_sequenceNextNode_Nullable;
```
It may be changed when merges happens, maybe we can make it deterministic? For example, always choose smallest `value` comparing as strings.